File: daq-server/tools/postprocessing/webviewer/backend/lc_tare.py
"""Load-cell tares recorded during a run, and the tared series derived from them.

Elodin records ABSOLUTE force, always and forever — `LC<n>_Cal.CH<c>.force_kg` means the same
thing in every run ever archived, and this feature did not change that. A tare is display state:
the backend subtracts it on the way to the browser and appends what it did to
`<run_dir>/lc_tare.jsonl`, so what the operator was looking at stays reconstructible afterwards.

This module replays that record. It exposes the reconstruction as a synthetic component,
`…force_kg_tared`, rather than as a toggle on the real one. A toggle would be a query parameter
threaded through three independent read paths (series_json, long_csv_rows, wide_csv_rows), and
missing one means the exported CSV disagrees with the plot the operator was looking at. As a
component name there is a single choke point in load_series, and series, both CSV shapes and the
download all follow for free.

The sidecar is append-only, one JSON object per line:

    {"entity":"LC2_Cal.CH1","uid":4201,"event":"set","offsetKg":20.13,
     "adcAtTare":8412331,"setAtMs":...,"appliedAtMs":1757800123456}

`appliedAtMs` is the instant the published stream changed, not when a button was pressed — the
backend writes the line from the same code that changes the subtraction. `event` is "set" (a new
tare), "recal" (a re-fit moved the offset with no operator action) or "clear". A recal line is
load-bearing: ignore it and every reconstruction is wrong from the re-fit onward.

There is no terminal line at session stop, so a tare with no following "clear" was in effect to
the end of the run.
"""

from __future__ import annotations

import json
from pathlib import Path

import numpy as np

from . import config

#: The component this module synthesises, and the real one it derives from.
TARED_SUFFIX = ".force_kg_tared"
GROSS_FIELD = "force_kg"


def sidecar_path(run_id: str) -> Path:
    """Inside the run dir, beside the DB — not a sibling like <run_id>.toml.

    That file is a sibling only because it is written before elodin-db creates the directory.
    This one is written mid-run when the directory certainly exists, and living inside it means
    the run's own deletion takes it too, with no orphan class to clean up.
    """
    return config.ELODIN_DIR / run_id / "lc_tare.jsonl"


def load(run_id: str) -> dict[str, list[tuple[float, float]]]:
    """entity -> [(applied_at_seconds, offset_kg), ...] ascending.

    Missing or malformed → {}. A viewer must never fail to open a run over its metadata, and a
    partially-written last line is expected: the file is appended to while the run is live.
    """
    p = sidecar_path(run_id)
    try:
        text = p.read_text()
    except OSError:
        return {}

    events: dict[str, list[tuple[float, float]]] = {}
    for line in text.splitlines():
        line = line.strip()
        if not line:
            continue
        try:
            rec = json.loads(line)
        except json.JSONDecodeError:
            # A torn final line is normal mid-run; earlier lines stay usable.
            continue
        entity = rec.get("entity")
        applied = rec.get("appliedAtMs")
        if not isinstance(entity, str) or not isinstance(applied, (int, float)):
            continue
        offset = 0.0 if rec.get("event") == "clear" else rec.get("offsetKg")
        if not isinstance(offset, (int, float)) or not np.isfinite(float(offset)):
            continue
        events.setdefault(entity, []).append((float(applied) / 1000.0, float(offset)))

    for ev in events.values():
        ev.sort(key=lambda e: e[0])
    return events


def tared_components(run_id: str, components: list[dict]) -> list[dict]:
    """The synthetic component entries to add to an index, one per tared LC channel.

    Only channels the sidecar actually names get one. Synthesising a tared twin for every load
    cell would make an untared run indistinguishable from a tared one whose offsets happened to
    be zero — the toggle would be lying about whether a tare existed.
    """
    events = load(run_id)
    if not events:
        return []
    out = []
    for comp in components:
        if comp.get("field") != GROSS_FIELD:
            continue
        entity = comp.get("entity", "")
        if entity not in events:
            continue
        d = dict(comp)
        d["name"] = f"{entity}{TARED_SUFFIX}"
        d["field"] = "force_kg_tared"
        out.append(d)
    return out


def apply(run_id: str, entity: str, t: np.ndarray, v: np.ndarray) -> np.ndarray:
    """Subtract the offset that was in effect at each sample time.

    A step function, not one offset across the whole array: a tare set partway through a run must
    leave everything before it absolute, or the pre-tare portion of every run reads wrong.
    """
    events = load(run_id).get(entity)
    if not events or len(t) == 0:
        return v
    times = np.array([e[0] for e in events], dtype=float)
    offsets = np.array([e[1] for e in events], dtype=float)
    # side="right": a sample exactly at the applied instant already carried the new offset.
    idx = np.searchsorted(times, t, side="right") - 1
    applied = np.where(idx >= 0, offsets[np.clip(idx, 0, len(offsets) - 1)], 0.0)
    return v - applied

File: daq-server/diablo_server/backend/src/lc-tare.ts
/**
 * Load-cell tare: the display-only subtraction, and the per-run record of it.
 *
 * WHY THE SUBTRACTION LIVES HERE AND NOT IN C++
 * ---------------------------------------------
 * The calibration service pushes raw ADC into its CaptureWindow *before* it computes any
 * kilograms, which is what makes a calibration capture structurally immune to a standing tare —
 * an operator can tare a loaded tank and still calibrate in absolute weight afterwards. Moving
 * the subtraction into C++ would put a tared number in front of that boundary and destroy the
 * property. So Elodin keeps carrying gross `force_kg` forever, every archive stays comparable,
 * and the tare is applied on the way to the browser and nowhere else.
 *
 * WHY A PARALLEL COMPONENT AND NOT AN IN-PLACE EDIT
 * -------------------------------------------------
 * The calibration page must be able to read absolute kilograms — the operator types the true
 * weight of a known mass, and a tared readout beside that input is how a false point gets into
 * the fit. Mutating `force_kg` would leave that page no way to get gross. So this emits
 * `force_kg_tared` alongside the untouched original and the frontend picks which to render.
 *
 * This is the backend's first derived stream component; everything else it publishes is verbatim
 * from the packet. That is a deliberate exception, for the reason above.
 */

import fs from 'fs';
import path from 'path';
import type { ParsedSensorData } from './elodin-protocol.js';
import { tarePath } from './routes/calibration-profiles.js';

/** One channel's tare as the calibration service wrote it. */
interface TareRecord {
  uid: number;
  entity: string;
  adc_at_tare: number;
  offset_kg: number;
  set_at_ms: number;
  curve_fp: number;
}

export interface TareEntry {
  uid: number;
  offsetKg: number;
  adcAtTare: number;
  setAtMs: number;
}

/** entity (e.g. "LC2_Cal.CH1") -> tare. Keyed on the string C++ writes; never re-derived here. */
type TareMap = Map<string, TareEntry>;

const COMPONENT_GROSS = 'force_kg';
export const COMPONENT_TARED = 'force_kg_tared';

let _cache: { mtimeMs: number; map: TareMap } | null = null;
/** The map currently applied to the stream. Survives a torn read; see loadTareMap(). */
let _held: TareMap = new Map();
let _dbDir: string | null = null;

/** Where the per-run record goes. Null between sessions — see appendRunRecord(). */
export function setRunDir(dbDir: string | null): void {
  _dbDir = dbDir;
}

/**
 * Forget everything. Called at session stop and on an Elodin reconnect: without it a stale
 * offset from the previous run is applied to the next run's first poll interval, which is
 * exactly the window in which nobody is looking closely yet.
 */
export function resetTareState(): void {
  _cache = null;
  _held = new Map();
}

function parseTareFile(text: string): TareMap {
  const map: TareMap = new Map();
  const root = JSON.parse(text) as { tares?: TareRecord[] };
  if (!root || !Array.isArray(root.tares)) throw new Error('no tares array');
  for (const t of root.tares) {
    if (!t || typeof t.entity !== 'string' || !t.entity) continue;
    // A non-finite offset is refused rather than clamped to 0: it means the service wrote
    // something it should not have, and silently treating it as "no tare" would hide that.
    if (!Number.isFinite(t.offset_kg)) continue;
    map.set(t.entity, {
      uid: Number(t.uid) || 0,
      offsetKg: t.offset_kg,
      adcAtTare: Number.isFinite(t.adc_at_tare) ? t.adc_at_tare : 0,
      setAtMs: Number.isFinite(t.set_at_ms) ? t.set_at_ms : 0,
    });
  }
  return map;
}

/**
 * The live tare map, reloaded when the file's mtime moves.
 *
 * Three outcomes, deliberately distinguished by error code rather than a bare catch:
 *   - file absent (ENOENT)  -> empty map. This is the normal state after a session-start clear.
 *   - file unreadable/torn  -> KEEP the previous map. The service writes tmp+rename, but a
 *                              reader can still lose a race; clearing on a transient parse
 *                              failure would make every LC plot jump by the tare amount for one
 *                              poll interval and then jump back.
 *   - file parsed           -> adopt it, and append any change to the run record.
 */
export function loadTareMap(): TareMap {
  let stat: fs.Stats;
  try {
    stat = fs.statSync(tarePath());
  } catch (e) {
    if ((e as NodeJS.ErrnoException).code === 'ENOENT') {
      if (_held.size > 0) {
        recordChanges(_held, new Map());
        _held = new Map();
      }
      _cache = null;
      return _held;
    }
    return _held; // permissions, I/O — hold what we have
  }

  if (_cache && _cache.mtimeMs === stat.mtimeMs) return _cache.map;

  let next: TareMap;
  try {
    next = parseTareFile(fs.readFileSync(tarePath(), 'utf8'));
  } catch {
    return _held; // torn or malformed — keep the last good map, do NOT clear
  }

  recordChanges(_held, next);
  _held = next;
  _cache = { mtimeMs: stat.mtimeMs, map: next };
  return next;
}

/** The offset to subtract for a cal entity. 0 when untared, absent, or non-finite. */
export function tareOffsetKg(entity: string): number {
  const t = loadTareMap().get(entity);
  if (!t) return 0;
  return Number.isFinite(t.offsetKg) ? t.offsetKg : 0;
}

/**
 * Add `force_kg_tared` for every calibrated load-cell point in `parsed`.
 *
 * Emits unconditionally, with a 0 offset when untared: a component that appears only while a
 * tare is set renders identically to a dead stream, and "is that channel untared or is the board
 * gone?" is not a question to ask an operator mid-procedure.
 *
 * Never touches `force_kg`, and never touches the `raw_adc`/`raw_adc_counts` echoes that ride in
 * the same array — those are the raw truth and feed calibration capture.
 */
export function expandWithTare(parsed: ParsedSensorData[]): ParsedSensorData[] {
  let out: ParsedSensorData[] | null = null;
  for (const p of parsed) {
    if (p.component !== COMPONENT_GROSS) continue;
    const offset = tareOffsetKg(p.entity);
    const value = p.value - (Number.isFinite(offset) ? offset : 0);
    // Guard again after the arithmetic: this value is produced downstream of the protocol
    // layer's finite check, so a NaN here would be dropped by the caller's own guard and the
    // entire tared series would vanish from the plot with nothing logged anywhere.
    if (!Number.isFinite(value)) continue;
    if (!out) out = [...parsed];
    out.push({ entity: p.entity, component: COMPONENT_TARED, value, timestamp: p.timestamp });
  }
  return out ?? parsed;
}

/**
 * Append what changed to <dbDir>/lc_tare.jsonl.
 *
 * The record is written by whoever applies the subtraction, at the instant the applied value
 * changes — not when the operator clicks, and not when C++ captures. Two reasons. The browser
 * can close between the command and the response while the tare still applies, which would
 * leave a run whose archive cannot be reconstructed for a tare that demonstrably happened. And
 * a re-cal rewrites offset_kg with no operator action at all, so there is no click to hang a
 * write off; this path covers it for free.
 */
function recordChanges(prev: TareMap, next: TareMap): void {
  if (!_dbDir) return; // a tare with no active session still applies; it just is not recorded
  const lines: string[] = [];
  const at = Date.now();

  for (const [entity, t] of next) {
    const before = prev.get(entity);
    if (before && before.offsetKg === t.offsetKg && before.adcAtTare === t.adcAtTare) continue;
    // Same code, different offset means the curve moved under a standing tare, not a new tare.
    const event = !before ? 'set' : before.adcAtTare === t.adcAtTare ? 'recal' : 'set';
    lines.push(
      JSON.stringify({
        entity, uid: t.uid, event, offsetKg: t.offsetKg,
        adcAtTare: t.adcAtTare, setAtMs: t.setAtMs, appliedAtMs: at,
      }),
    );
  }
  for (const [entity, t] of prev) {
    if (next.has(entity)) continue;
    lines.push(
      JSON.stringify({
        entity, uid: t.uid, event: 'clear', offsetKg: 0,
        adcAtTare: t.adcAtTare, setAtMs: t.setAtMs, appliedAtMs: at,
      }),
    );
  }
  if (lines.length === 0) return;

  try {
    fs.mkdirSync(_dbDir, { recursive: true });
    // Append, never rewrite: each line is a thing that happened. There is no terminal line at
    // session stop, so a missing trailing 'clear' unambiguously means "in effect to end of run".
    fs.appendFileSync(path.join(_dbDir, 'lc_tare.jsonl'), lines.join('\n') + '\n');
  } catch (e) {
    console.warn('[LcTare] could not append the run record:', e);
  }
}

/** The live tares, for the status endpoint the calibration UI polls. */
export function currentTares(): Array<TareEntry & { entity: string }> {
  return [...loadTareMap()].map(([entity, t]) => ({ entity, ...t }));
}
